### PR TITLE
added moment_p sampling

### PR DESCRIPTION
**Pull request by Hellisotherpeople**
# Moment-P Sampling Implementation for Hugging Face Transformers

## Overview

This implementation adds a new sampling method called "Moment-P" to the Hugging Face Transformers library. Moment-P is a novel sampling technique that filters tokens based on the moment (sum of probabilities raised to a power) of the probability distribution. It uses the whole probability distribution and enables very high temperature generations with coherence. A paper is incoming. 

## What is Moment-P Sampling?

Moment-P sampling filters tokens by calculating a threshold based on the moment of the probability distribution. The moment is calculated as:
```
moment = Σ(p_i^exponent)
```

Where:
- `p_i` is the probability of token i
- `exponent` is a configurable parameter (default: 2.0)

Tokens are kept until their cumulative moment exceeds `total_moment * alpha`.

### Key Parameters

- **moment_p_exponent** (float, default=2.0): The exponent to which probabilities are raised
- **moment_p_alpha** (float, default=1.0): Scaling factor for the threshold

### Behavior with Different Exponents

- **exponent=1**: Linear moment, truncate all tokens
- **exponent=2** (default): Second moment, emphasizes variance-like behavior
- **exponent>2**: Higher emphasis on the tail of the distribution, lower truncation values

## Implementation Details

The implementation consists of:

1. **MomentPLogitsWarper** class in `logits_process.py`
2. Configuration parameters in `configuration_utils.py`
3. Integration in `utils.py` for the generation pipeline
4. Export in `__init__.py`

### Basic Usage

```python
from transformers import AutoTokenizer, AutoModelForCausalLM

model = AutoModelForCausalLM.from_pretrained("distilbert/distilgpt2")
tokenizer = AutoTokenizer.from_pretrained("distilbert/distilgpt2")

inputs = tokenizer("The future of AI is", return_tensors="pt")

# Generate with moment_p sampling
output = model.generate(
    **inputs,
    do_sample=True,
    moment_p_exponent=2.0,
    moment_p_alpha=1.0,
    max_new_tokens=50
)

print(tokenizer.decode(output[0], skip_special_tokens=True))
```

### Advanced Usage with Different Parameters

```python
# More selective filtering with lower alpha
output = model.generate(
    **inputs,
    do_sample=True,
    moment_p_exponent=2.0,
    moment_p_alpha=0.5,  # More selective
    max_new_tokens=50
)

# Higher exponent for more tail emphasis
output = model.generate(
    **inputs,
    do_sample=True,
    moment_p_exponent=3.0,  # Higher moment
    moment_p_alpha=1.0,
    max_new_tokens=50
)
```

## Mathematical Foundation

The moment-based approach allows for flexible control over the token selection process:

1. **Calculate probabilities**: `probs = softmax(logits / temperature)`
2. **Calculate total moment**: `total_moment = Σ(probs^exponent)`
3. **Set threshold**: `threshold = total_moment * alpha`

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

@gante 

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Hellisotherpeople** [comment]
Some changes that shouldn't have made it in made it in. I will fix this and hopefully most tests pass.

**Comment by ArthurZucker** [comment]
cc @gante

**Comment by gante** [comment]
Hi @Hellisotherpeople 👋 Thank you for opening the PR, and for the detailed description!

`transformers` is a massive library at the moment, and we're being much more selective when adding model-agnostic features. We want to prevent the number of flags from growing uncontrollably, and instead improve the UX and quality of the methods we enable inside the library (including documentation and tests).

This doesn't mean we don't want to support techniques like yours 🤗 We've been adding features to enable both good UX on popular methods and support of new goodies like your logits processor. We're currently following a two-step approach to add new decoding methods / logits processors / ... (anything `generate`-related) to the main codebase:
1. Any modification to `generate` can be added as a custom generation method ([docs](https://huggingface.co/docs/transformers/generation_strategies#custom-decoding-methods)). This should make your method easily accessible to ALL `transformers` users. In this specific case, you want to initialize a new logits processor before calling the original `generate`, it should be quite simple to build.
2. If a given technique becomes popular (e.g. amazing paper results, custom generation method with many downloads, hot social media discussion), then we add it to `transformers`. In other words, merge this PR. (The code itself looks good -- with a few more checks in the test, it would be ready to be merged.)

I hope you understand our decision to push back on PRs with new `generate`-related features. Let me know if you'd like help with the custom generation repo, or with anything else :)

**Comment by Hellisotherpeople** [comment]
Hi all, let's look at this through the lens of our shared success with min-p, which I helped co-authored and delivered the [ICLR 2025 Oral presentation for](https://iclr.cc/virtual/2025/oral/31888). **The merge for min_p**, while difficult, **was the non-negotiable prerequisite for its success.**. 

PLEASE WATCH that oral presentation video and read the slides as I make the case for why HF is a leader in LLM sampling innovation today in it! 

Here is the core of the issue:

1. Adoption Follows Access: You mentioned concerns about potential utility or bloat. **However, a sampler's value cannot be judged before it's been released.** The community and researchers **need** access via a widely used library like transformers to discover its value.

2. Research Dependency: Our lm-evaluation-harness experiments, which are the backbone of our upcoming paper, are blocked without this merge. **transformers support is not optional for our research validation; it's a critical dependency. The min_p paper would not exist if you had not eventually merged that PR.**. There is NO support for custom_generate in lm-evaluation-harness and it is the core for running good evaluations.

We believe that transformers adds immense value by being a hub for research. After all, you have existing samplers (e.g., typicality sampling) that aren't widely used but were added to explore the research space. Is this principle no longer active? Grandfathering in unused samplers and not letting new, potentially highly novel ones in is a terrible strategy for any library that wants to keep being utilized for SOTA research. 

We intend to push this method to other frameworks like vllm and sglang regardless, as we need it for our work. This means the community will find and use this method... eventually anyway... (it took min_p 2+ years to get a paper) The only question is whether transformers will be credited as a go-to library for fostering such innovation.

If this PR is rejected on principle, **we will be forced to note in our paper and subsequent public discussions that Hugging Face's contribution policy has become a roadblock for sampler research, and to direct researchers to use the other platforms where the technique was implemented.**. 

In particular, you will find that very soon in a world of intense competition between sglang, vllm, TGI for dominance of LLM inference space - that one of the main competitive differentiators will be in access to more advanced sampling algorithms. This is a space where billions of dollars of value is being created. 

We genuinely hope to avoid this and instead continue highlighting Hugging Face as a collaborative and essential partner in AI research.

**Comment by gante** [comment]
@Hellisotherpeople, your comment assumes `moment_p` wouldn't be immediately accessible to everyone, which isn't the case at all 🤗  If anything, the process I suggest removes blockers from the HF side: now anyone can add a variation to text generation with zero oversight from a HF team member, as opposed to needing to go through a PR review.

The suggested workflow, from a usage perspective, simply changes

```py
model.generate(**model_inputs, moment_p_exponent=2.0, moment_p_alpha=1.0)
```

to

```py
model.generate(**model_inputs, moment_p_exponent=2.0, moment_p_alpha=1.0, custom_generate="your/repo", trust_remote_code=True)
```

Moreover, in the dedicated repo, you can add a landing page with as much information as you want, as opposed to in `transformers`, where docs would be limited to a handful of lines.

If you find there are still blockers to this usage path, including outside `transformers`, let me know -- I'm here to remove them :)

__________________________

We need to understand that we have new variations to `generate` coming out every month, and each one we add to `transformers` brings new flags, tests, documentation, and a long-term maintenance burden to the library. Moreover, they often fail the test of time and, contrary to models (where we only use one at a time), they can be used with each other. If we add them all to `transformers`, it will result in a terrible user experience (too many options) and poorly maintained code. A park of dead projects, like our `examples` folder.

Precisely because we want to foster a researcher-friendly environment, we've added a process that can scale well and has *fewer* blockers: code on the hub to modify `generate`, no gatekeeping by HF members on `transformers` PRs, and that is immediately usable by anyone who has `transformers` installed.

**Comment by gante** [comment]
@Hellisotherpeople, it is my fault for shifting the goalposts (see my explanation below). I will review your PR to include it in the library. 

We've reviewed our library-wide priorities, and thus acceptance criteria for `generate`-related topics. Stopping criteria & logits processors are to be accepted (unless they're overly complex). We're still trying to figure out how to best balance maintenance with contributions, please bear with us. At the moment, our rationale is to exclude anything that DOESN'T fit any of this criteria:
- Helps reducing the complexity of the library as a whole
- Fixes a critical bug
- Feature that can be reused with continuous batching, i.e. logits processors and stopping criteria
- (Unlikely event of a decoding method with ungodly performance gains)

File: src/transformers/generation/configuration_utils.py [hunk 1]
```diff
@@ -421,6 +429,8 @@ def __init__(self, **kwargs):
         self.top_k = kwargs.pop("top_k", 50)
         self.top_p = kwargs.pop("top_p", 1.0)
         self.min_p = kwargs.pop("min_p", None)
+        self.moment_p_exponent = kwargs.pop("moment_p_exponent", 2.0)
+        self.moment_p_alpha = kwargs.pop("moment_p_alpha", 1.0)
```
Comment (gante) [inline src/transformers/generation/configuration_utils.py:433]: New arguments should default to `None`, so that we can distinguish e.g. a user-defined `moment_p_exponent=2.0` from the default value.

It seems equivalent now, but non-None defaults often bring issues in the log run

File: src/transformers/generation/configuration_utils.py [hunk 2]
```diff
@@ -211,6 +211,14 @@ class GenerationConfig(PushToHubMixin):
...
             value between 0 and 1. Typical values are in the 0.01-0.2 range, comparably selective as setting `top_p` in
             the 0.99-0.8 range (use the opposite of normal `top_p` values).
+        moment_p_exponent (`float`, *optional*, defaults to 2.0):
+            The exponent to which probabilities are raised when calculating the moment. Higher values emphasize the
+            tail of the distribution more and reduce the truncation threshold. Values closer to 1 increase the threshold
+            to approach 1 as this value approaches 1. The default value represents the "P-Less" method's second moment,
+            i.e. `moment_p_exponent=2.0`.
+        moment_p_alpha (`float`, *optional*, defaults to 1.0):
+            Scaling factor applied to the moment threshold. Values < 1 make filtering more selective, while values > 1
+            make it less selective.
```
Comment (gante) [inline src/transformers/generation/configuration_utils.py:221]: Suggestion: if `moment_p` needs these two parameters to be correctly set, let's consider a dictionary (or a tuple). That way, users can immediately know that the attributes are used together.

I am aware other techniques have separate args, due to poor foresight 😢

**Review by gante — COMMENTED** [review]
@Hellisotherpeople Other than a missing integration test* and the comment below, LGTM :)

*Integration test example: https://github.com/huggingface/transformers/blob/88e946e06233b6e470f44443b71ef45829ef5d30/tests/generation/test_utils.py#L3970
